diff --git a/netfilter_nft/nftnlrdr_misc.c b/netfilter_nft/nftnlrdr_misc.c
index 7f227ee..da65be0 100644
--- a/netfilter_nft/nftnlrdr_misc.c
+++ b/netfilter_nft/nftnlrdr_misc.c
@@ -305,7 +305,7 @@ parse_rule_nat(struct nftnl_expr *e, rule_t *r)
 	reg_val_ptr = get_reg_val_ptr(r, addr_min_reg);
 	if (reg_val_ptr != NULL) {
 		r->nat_addr = (in_addr_t)*reg_val_ptr;
-		if (proto_min_reg == NFT_REG_1) {
+		if (proto_min_reg == NFT_REG_1 || proto_min_reg == NFT_REG_2) {
 			r->nat_port = proto_min_val;
 		}
 	} else {
